octoprint_printhistory/parser.py

```diff
@@ -0,0 +1,294 @@
+        settings = settings.replace('{"global_quality": "', "")
+        settings = settings.replace('"}', "")
+        config = ConfigParser.RawConfigParser(allow_no_value=True)
+        config.readfp(io.BytesIO(settings))
+        try:
+            for section in ["values", "metadata"]:
+                for option in config.options(section):
+                    parameters.update({option: config.get(section, option)})
+        except:
+            pass
+        return parameters
+
+
+class Slic3rParser(BaseParser):
+
+    def __init__(self):
+        self.name = "slic3r"
+        self.version = None
+        self.is_parameter = re.compile(r"^[\w _]+={1}")
+
+    def detect(self, file):
+        detected = False
+        # on the first line
+        # ; generated by Slic3r 1.2.9 on 2017-01-30 at 21:53:46
+        line = file.readline()
+        file.seek(0)
+        if re.search(r"Slic3r", line):
+            self.version = VERSION_REGEX.search(line).group(0)
+            detected = True
+        return detected
+
+    def parse_header(self, file):
+        parameters = {}
+        for line in file:
+            if line.startswith(";"):
+                line = line.split(";")[1].strip()
+                if self.is_parameter.match(line):
+                    splitted = line.split("=")
+                    param = splitted[0].strip()
+                    value = splitted[1].strip()
+                    parameters.update({param: value})
+            elif line == "\n":
+                continue
+            else:
+                break
+        file.seek(0)
+        return parameters
+
+    def parse_bottom(self, file):
+        parameters = {}
+        for line in reverse_readline(file):
+            if line.startswith(";"):
+                line = line.split(";")[1].strip()
+                if self.is_parameter.match(line):
+                    splitted = line.split("=")
+                    param = splitted[0].strip()
+                    value = splitted[1].strip()
+                    parameters.update({param: value})
+            elif line == "\n":
+                continue
+            else:
+                break
+        return parameters
+
+
+class Simplify3DParser(BaseParser):
+
+    def __init__(self):
+        self.name = "simplify3d"
+        self.version = None
+        # checks if string starts with a word followed by a comma
+        self.is_parameter = re.compile(r"^\w+,{1}")
+        self.is_bparameter = re.compile(r"^[\w ]+:{1}")
+
+    def detect(self, file):
+        detected = False
+        # on the first line
+        # ; G-Code generated by Simplify3D(R) Version 3.1.0
+        line = file.readline()
+        file.seek(0)
+        if re.search(r"Simplify3D\(R\)", line):
+            self.version = VERSION_REGEX.search(line).group(0)
+            detected = True
+        return detected
+
+    def parse_header(self, file):
+        parameters = {}
+        for line in file:
```

Comment:
`file` is a **TERRIBLE** variable name since it's shadowing python builtin: http://stackoverflow.com/a/24942363/699864  Even GitHub highlites it's differently.

Please in the name of a mother's love use IDE already.